### ISSUE TITLE
Game Mechanics/Rules

### ISSUE THREAD

**Issue by hzoo** [hzoo/shengji#3]
Listing out the API for the specifics of the game itself (basically the interfaces)

> Also need to define terms consistently.
#### Random Rules
- Drawing: each player draws one at a time in counter-clockwise order.
- Trump Level/Rank: equal to the rank of the Attackers in that round.
- Trump Suit: determined at the beginning of each round.
#### Terms

**Level**: the score/rank of a team or player.
**Trick**: a single unit/round of play (each player plays their cards).
**Leading Player**: describing the player that leads/starts/begins a trick.
**Points**: certain cards (5,10,K) are worth points.
**Kitty**: describes the bottom pile of cards.
**Trump**: a card that is elevated above its normal rank.
**Trump Suit**: these cards outrank plain (non-trump) suits.
**Dynamic Trump**: the trump suit can change.
**Attackers**: one of the two sides in the game - goal is to obtain points.
**Defenders**: the other side in the game  - goal is to prevent the other team from gaining points.
**Dealer**: sometimes known as "master" or "king" - a special role on the attacking side.

Initial thoughts...
- [ ] validateLeadingPlayerHand: Boolean - check if the leading players played cards are possible in a trick.
- [ ] validateFollowingPlayerHand: Boolean - check if the other's players played cards are possible in a trick.
- setupDecks, shuffle, dealCards
#### Round
- setTrump
- nextTrick
- play
- endTrick
- endGame
- [ ] getNextDealer: Number
- [ ] getNewPlayerLevel: Number
- [ ] getNewPlayerLevels: Array<Number>
#### Trick
- play
- winner
- points
- pointCards
##### Methods
- [x] calcPointsForSingleCard: Number - calculate the point of a single card.
- [x] calcPoints: Number - calculate the points of an array of cards.
- [x] maxPoints: Number - get the maximum points for a given # of packs
- [x] calcKittyPoints: Number - calculate the points of the bottom cards.
- [x] numDecks: Number - get the number of decks for a given # of players
- [ ] isStronger: Boolean - if a hand > another hand (need bit more logic to check tractor pairs)
- [x] isAPointCard: Boolean
- [x] isTrump: Boolean
##### Config
- [x] cardValue - in order to calculate the winner
- [x] cardSuit - suits
- [x] pointIncreaseThreshold - percentage of total points to increase in level
- [x] pointCards - which cards are points
## Links

https://www.quora.com/What-are-some-strategies-for-the-card-game-Finding-Friends

**Comment by kumquatexpress** [comment]
I was thinking cards would be a model that represented playable "moves", so the model would contain some number of actual cards with suits, as well as the overall "strength" of the play dependent on who played it for comparisons within a trick.

Having an object hold just a value and suit doesn't seem necessary.

**Comment by hzoo** [comment]
Yeah I actually had to create some weird model for [comparisons](https://github.com/hzoo/Tuolaji-Nonoba/blob/master/Serverside%20Code/Game.cs#L307). What a crazy function - don't know if I can read it still. I made a tuolaji/shuai string basically. like AAKK would be A2K2 I think. And I tried to compute the highest type of hand it would be. It seems messy but it worked at the time - the suits don't matter unless its trump (the object is just to verify correct suits etc).

Haven't gotten around to looking at it too much yet. Ok we'll need some function to create a players hand to some format we can use to compare hands.
